File: leetcode/weekly-contests/contest-188-virtual/1443-minimum-time-to-collect-all-apples-in-a-tree/dfs.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        nodes = defaultdict(list)
        for ele in edges :
            nodes[ele[0]].append(ele[1])
            nodes[ele[1]].append(ele[0])
        visited = [False]*n
        
        
        def helper(root , currentCost) :
            if(visited[root]) :
                return 0
            visited[root] = True
            childrenCost = 0
            for child in nodes[root] :
                childrenCost += helper(child , 2)
            if(childrenCost == 0 and hasApple[root] == False) :
                return 0
            return childrenCost + currentCost
    

        return helper(0 , 0)

File: leetcode/weekly-contests/contest-188-virtual/1443-minimum-time-to-collect-all-apples-in-a-tree/dfs.cpp
class Solution {
    int ans = 0 ;
    unordered_map<int , vector<int>>nodes ;
    vector<bool>visited ;
public:
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) ;
    bool helper(int n , vector<bool>& hasApples , int root) ;
};

bool Solution :: helper(int n , vector<bool>& hasApple , int root)
{
    if(visited[root])
        return false ;
    visited[root] = true ;
    bool increment = hasApple[root] ;
    for(auto child : nodes[root])
    {
        bool cond = helper(n , hasApple , child) ;
        if(cond) ans += 2 ;
        increment = cond or increment ;
    }
    
    return increment ;
}

int Solution :: minTime(int n , vector<vector<int>>& edges , vector<bool>& hasApple)
{
    visited.resize(n , false) ;
    for(auto ele : edges)
    {
        nodes[ele[0]].push_back(ele[1]) ;
        nodes[ele[1]].push_back(ele[0]) ;
    }
    bool temp = helper(n , hasApple , 0) ;
    return ans ;
}
